### PR TITLE
CI: Build Embedded Boundaries

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

- [x] Add coverage for EB in CI
- [x] fix DPC++ single precision build
- [x] ~~fix `LinearSolvers` array bounds w/ GCC 7.5~~ -> false positive

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
Do we want some tests that do not use EB?  It's possible to have errors that appear only when EB is not enabled.

**Comment by ax3l** [comment]
> Do we want some tests that do not use EB? It's possible to have errors that appear only when EB is not enabled.

Good idea - in WarpX I add a single CI test that builds a "minimal" build - disabling as much features as possible.
That way, we can cover that we can truly disable features at compile-time.

**Comment by ax3l** [comment]
@WeiqunZhang can you take a look at the `-Werror=array-bounds` error in `Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp`?
```
Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp
In file included from /home/runner/work/amrex/amrex/Src/Base/AMReX_ArrayLim.H:9:0,
                 from /home/runner/work/amrex/amrex/Src/Base/AMReX_Box.H:7,
                 from /home/runner/work/amrex/amrex/Src/Base/AMReX_FabFactory.H:6,
                 from /home/runner/work/amrex/amrex/Src/EB/AMReX_EBFabFactory.H:5,
                 from /home/runner/work/amrex/amrex/Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.H:5,
                 from /home/runner/work/amrex/amrex/Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp:2:
/home/runner/work/amrex/amrex/Src/Base/AMReX_Array.H: In member function ‘virtual void amrex::MLEBABecLap::applyBC(int, int, amrex::MultiFab&, amrex::MLLinOp::BCMode, amrex::MLLinOp::StateMode, const amrex::MLMGBndry*, bool) const’:
/home/runner/work/amrex/amrex/Src/Base/AMReX_Array.H:114:23: error: array subscript is above array bounds [-Werror=array-bounds]

**Comment by WeiqunZhang** [comment]
I suspect this is false positive.  GCC's `-Werror=array-bounds` has a lot of false positives.  I could not reproduce it on my desktop with gcc 9.3.0.  May have to use gcc 7.5.

**Comment by ax3l** [comment]
Thanks for the check, see it now as well in the other entries.
I would say for the sake of compiler coverage and since we have also newer GCCs in CI, I would add `-Wno-array-bounds` to the one GCC 7.5 build that has the false positive.

Alternatively, I can also just disable EB in that one build again, to keep the very first build small and minimal. I pushed that for now.

What do you think?

**Comment by WeiqunZhang** [comment]
Sounds good.  When I get time, I will install gcc 7.5 and figure out what it complains about.

**Comment by ax3l** [comment]
Looking where this is coming from:
```
36/38 Test #36: LinearSolvers_NodalPoisson ..........***Failed    0.72 sec
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (21.08.0) initialized
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
SIGABRT
See Backtrace.0 file for details
Abort(6) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 6) - process 0
```

**Comment by WeiqunZhang** [comment]
Possible reason is that the non-EB test does work when EB is enabled.

**Comment by WeiqunZhang** [comment]
I can push a fix to this PR.

**Comment by ax3l** [comment]
I think the bad cast is here:
https://github.com/AMReX-Codes/amrex/blob/0ca401daa57fd80e7ae7a12eb5fd28b9d10ce11b/Src/EB/AMReX_algoim.cpp#L25

(With the last push it might now be one call earlier.)

**Comment by WeiqunZhang** [comment]
That's right.  I added a test before calling that function and that should also work.  The CI now fails for a different reason.

**Comment by ax3l** [comment]
Is it possible the EB unit tests are pretty large in terms of runtime? I see the CI on ICC crunching on it for 20+ min

**Comment by WeiqunZhang** [comment]
Yes, some EB tests are too large for CI.

**Comment by ax3l** [comment]
Those are the biggest runtime eaters:
```
30/38 Test #30: CNS_Combustor .......................   Passed  145.31 sec
      Start 31: CNS_Pulse
32/38 Test #32: CNS_ShockRef ........................   Passed  4259.57 sec
      Start 33: CNS_Sod
33/38 Test #33: CNS_Sod .............................   Passed  112.03 sec
```

@WeiqunZhang can we just reduce their resolution and steps for the checked in inputs?
If they need to be large they might be better candidates for tutorials than the tests in the AMReX repo?

**Comment by WeiqunZhang** [comment]
Yes, we can try to reduce the size or exclude some inputs files.

I also want to look at the EB coverage.  We already have some CI tests for EB in development.  I think we are converting too many non-EB tests to EB now.

**Comment by ax3l** [comment]
For this PR to finish, shall I exclude all `CNS*` tests for now and we reduce in a follow-up?

**Comment by ax3l** [comment]
Excluded the `CNS_*` test in the last commit from CI.
Let's re-enable them in #2237 once we reduce the inputs.

**Comment by ax3l** [comment]
@WeiqunZhang ready for merge :)

**Comment by WeiqunZhang** [comment]
In AMReX, the EB codes are either put inside `#ifdef AMREX_USE_EB` or excluded in the build so that non-EB builds could work.  The issue in this PR is now the majority of the tests has enabled EB.  So, the CI tests may not be able to catch non-EB build errors.  Currently in the development branch, we are already testing EB builds in 4 GNU make CI tests with a combination of 2d, 3d, mpi, omp, hip, and cuda.  We want to add more coverage especially with cmake, but I think we should still leave most of the CI tests as non-EB builds.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -143,6 +153,8 @@ jobs:
             -DCMAKE_Fortran_COMPILER=$(which gfortran-10)
         make -j 2
 
+        ctest --output-on-failure -E Particles_GhostsAndVirtuals
```
Comment (ax3l) [inline .github/workflows/linux.yml:156]: Those should become addressed with #2247 (issue: #2243).
We can comment those in again once the linked PR is merged

**Comment by ax3l** [comment]
@WeiqunZhang green now :)

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**